hooks/pre-push

```diff
@@ -1,2 +1,13 @@
 #!/bin/bash -e
-for f in $0.d/*; do $f; done
+rust=0
+for file in $(git diff-index --name-only HEAD); do
```

Comment:
I'm not 100% sure what this command does. But let me think aloud here for a moment. I can think of the following cases:

1. `git push origin foo` -- new branch `foo`

In this case we are creating a remote branch and pushing all commits to it. We have no way to determine which commits are "new". We run tests.

2. `git push origin foo` -- append to existing branch `foo`

We need a list of all files changed in the commits we are appending. We conditionally run tests if code files have changed.

3. `git push -f origin foo` -- overwrite existing branch

We don't know which commits are "new". Do we run tests? I'm not sure. It is certainly nice to have an option to disable tests (i.e. `-f` == don't run tests). However, when doing things like squashing commits we need to specify `-f` **and** run tests. So I think we have to run tests.

4. CI/CD

This case is similar to the appending case above, but rather than appending we are managing the difference between two forks. So we need to know which commits differ between two revisions.

I'm honestly wondering if this problem is too hard to solve right now... Or conversely, if trying to run tests pre-push was the wrong decision. Thoughts?

Reply:
To run through my logic when I wrote this:

The command (`git diff-index --name-only HEAD`) lists all filenames that differ from HEAD in the current branch. (I got the command [here](https://stackoverflow.com/questions/3882838/whats-an-easy-way-to-detect-modified-files-in-a-git-workspace).) The `for` loop iterates through all these files and checks if they meet some criteria that would indicate we need to run full Rust tests:

- a `.rs` or `.toml` file was edited
- we are running on CI (Travis, in this case)

If any of them are met, we flip a status bit. That's all we need to determine we need to run tests, so the loop breaks.

We then check, for every test in the `pre-push.d` directory, if they are a `cargo-` test. If they are, the status bit must also be enabled to run them. Otherwise, the test is run regardless.

I suppose a lot of my logic is predicated upon the `git diff-index` command accurately telling us which files have changed. I think if that assumption is sound, a lot of these cases are adequately addressed with the current logic. (That said, I'll need to debug Travis tests... doesn't seem like they're force-run yet.)

I will also say that I like having pre-push tests (or more generally, having tests run locally before pushing), as long as we can make them get out of the way when not needed. Is your thought more about the fact we should change where they're run, or whether we should have git hooks tests at all?

Reply:
@mbestavros The problem is that you're only checking the last commit. But the user might be pushing more than one commit. So imagine a case like this:

Commit 1: change some code
Commit 2: change some docs

The existing code would look only at commit 2. Seeing no code changes, it will skip running tests. The problem is that both commits are new.

This leads to the four scenarios I listed above.

Reply:
I see now.

I suppose we want to compare the current branch with `upstream/master`, which may be difficult if an upstream has not been set up... I think if we can reliably determine those things, we can make accurate judgement calls for all cases above.

So the question of the hour: is there a command that will do that?

Reply:
I don't think there's a ready-made way to do this, no. There are also other problems.

First, writing tests that have to execute in multiple special environments can be challenging.

Second, pre-push hooks cannot be enabled by default. They are opt-in.

Third, opting in to pre-push hooks means developers have to setup special environments just for testing. This isn't developer friendly.

Fourth, tests currently run in the working tree: https://github.com/enarx/enarx/issues/214 .

Hence, I'm wondering if this was a mistake (which I will completely own). Perhaps we should just focus on the CI/CD problem and migrate our tests there.